metrics/repl.lag/lag.go

```diff
@@ -115,70 +130,122 @@ func (c *Lag) Help() blip.CollectorHelp {
+			case LAG_SOURCE_BLIP:
+				return c.prepareBlipHeartbeatLag(levelName, monitorID, monitorName, options)
+			default:
+				return nil, fmt.Errorf("invalid lag source: %s; valid values: auto, pfs, blip", source)
+			}
 		}
+	}
+
+	// -------------------------------------------------------------------------
+	// Auto source (default)
+	// -------------------------------------------------------------------------
+	var err error
+	if err = c.preparePFS(levelName); err == nil {
+		return nil, nil
+	}
+
+	if cleanup, err := c.prepareBlipHeartbeatLag(levelName, monitorID, monitorName, options); err == nil {
+		return cleanup, nil
+	}
+
+	return nil, fmt.Errorf("auto lag source failed, last error: %s", err)
+}
 
-		c.dropNoHeartbeat[level.Name] = !blip.Bool(dom.Options[OPT_REPORT_NO_HEARTBEAT])
-		c.dropNotAReplica[level.Name] = !blip.Bool(dom.Options[OPT_REPORT_NOT_A_REPLICA])
+func (c *Lag) prepareBlipHeartbeatLag(levelName string, monitorID string, planName string, options map[string]string) (func(), error) {
+	if c.lagReader == nil {
+		c.dropNoHeartbeat[levelName] = !blip.Bool(options[OPT_REPORT_NO_HEARTBEAT])
 
+		table := options[OPT_HEARTBEAT_TABLE]
+		if table == "" {
+			table = blip.DEFAULT_HEARTBEAT_TABLE
+		}
 		netLatency := 50 * time.Millisecond
-		if s, ok := dom.Options[OPT_NETWORK_LATENCY]; ok {
+		if s, ok := options[OPT_NETWORK_LATENCY]; ok {
 			n, err := strconv.Atoi(s)
 			if err != nil {
-				blip.Debug("%s: invalid network-latency: %s: %s (ignoring; using default 50 ms)", plan.MonitorId, s, err)
+				blip.Debug("%s: invalid network-latency: %s: %s (ignoring; using default 50 ms)", monitorID, s, err)
 			} else {
 				netLatency = time.Duration(n) * time.Millisecond
 			}
 		}
-
-		switch dom.Options[OPT_WRITER] {
-		case "", "blip": // "" == default == blip
-			if c.lagReader == nil {
-				// Only 1 reader per plan
-				c.lagReader = heartbeat.NewBlipReader(heartbeat.BlipReaderArgs{
-					MonitorId:  plan.MonitorId,
-					DB:         c.db,
-					Table:      table,
-					SourceId:   dom.Options[OPT_SOURCE_ID],
-					SourceRole: dom.Options[OPT_SOURCE_ROLE],
-					ReplCheck:  sqlutil.CleanObjectName(dom.Options[OPT_REPL_CHECK]), // @todo sanitize better
-					Waiter: heartbeat.SlowFastWaiter{
-						MonitorId:      plan.MonitorId,
-						NetworkLatency: netLatency,
-					},
-				})
-				go c.lagReader.Start()
-				blip.Debug("%s: started reader: %s/%s (network latency: %s)", plan.MonitorId, plan.Name, level.Name, netLatency)
-			}
-			c.atLevel[level.Name] = true
-		}
+		// Only 1 reader per plan
+		c.lagReader = heartbeat.NewBlipReader(heartbeat.BlipReaderArgs{
+			MonitorId:  monitorID,
+			DB:         c.db,
+			Table:      table,
+			SourceId:   options[OPT_HEARTBEAT_SOURCE_ID],
+			SourceRole: options[OPT_HEARTBEAT_SOURCE_ROLE],
+			ReplCheck:  c.replCheck,
+			Waiter: heartbeat.SlowFastWaiter{
+				MonitorId:      monitorID,
+				NetworkLatency: netLatency,
+			},
+		})
+		go c.lagReader.Start()
```

Comment:
no need to signal to the `Start` routine for a clean exit?

Reply:
`c.lagReader.Stop()` is explicitly called in the cleanup func to stop the reader before exit.